[WIP] 📖 ClusterClass patch proposal amendment

sbueringer opened the pull request:
Signed-off-by: Stefan Büringer buringerst@vmware.com

<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
This amendment introduces ClusterClass patches.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #5142

sbueringer commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+  Kind string `json:"kind"`
+ 
+  // MatchResources selects templates based on where they are referenced.
+  MatchResources PatchMatchResources `json:"matchResources"`
```
I wonder if something like `matchContext` would be better.

vincepri replied:
Resources sounds good to me as-is, given that it's a known Kubernetes concept

sbueringer replied:
Was just wondering because we're kind of matching **on** the context where those resources are used and not the resources themselves

sbueringer commented:
/retest

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +139,25 @@ For an end user, the UX to update the kubernetes version of the control plane an
 - Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
 
+#### Story 4 - Ability to customize individual clusters
+As a Cluster operator, I want to customize individual clusters simply by providing variables in the Cluster resource.
```
can we be more specific here about what is "small deviations" may be going through a real world example that illustrate how the different personas benefit from this level of customization?

sbueringer replied:
Added an example, please let me know if that clarifies it.

tl;dr is basically if we don't have variables you have to create new ClusterClasses whenever you want to set a single field in any of the referenced templates differently. Looking at the cluster-templates we have in our providers (e.g. CAPA/CAPV), there are a lot of those.

schrej commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -261,8 +468,20 @@ type LocalObjectTemplate struct {
+	  // VariableDefinitionClasses defined in the ClusterClass.
+	  // +optional
+	  Variables []VariableTopology `json:"variables,omitempty"`
     }
```
Have you considered to also allow Patches on individual clusters?

While the goal with ClusterClass obviously is to create clusters that are as homogenous as possible, it might be useful to be able to apply patches to individual clusters that weren't considered when creating the Class. Especially applying hotfixes to individual clusters would be a good use case for this, as it isn't possible to modify any of the cluster resources when they are managed by CAPI.

sbueringer replied:
I think this use case is pretty close to what could be achieved with optional patches (which is not part of the current iteration). The idea of optional patches is that you can opt-in into patches of the ClusterClass in the Cluster resources.

So the workflow for hotfixes would be:
* Add a new optional patch to the ClusterClass
* Enable the optional patch on a per-cluster basis

Imho we should try to avoid introducing the whole complexity of patches on the Cluster resource.

sbueringer replied:
I adjusted the note about optional patches a bit
```
**Note**: We are considering to add a field to allow defining optional patches. This would allow
adding optional patches in the ClusterClass and then activating them on a per-cluster basis via opt-in.
```

schrej commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+ 
+// VariableDefinitionClass defines a variable which can 
+// be configured in the Cluster topology and used in patches.
+type VariableDefinitionClass struct {
```
A default value could also be useful here, especially when variables are used in conjunction with patches. Especially since arrays can only be modified using `add`.

schrej replied:
Nevermind, that's done using the Schema.

schrej commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,34 +532,124 @@ type LocalObjectTemplate struct {
+    - schemas are mutable
+      - The current assumption is that we validate schema changes against existing clusters and block in case the changes are 
+        not compatible. We are considering to implement a clusterctl command to allow the user to simulate the impact of a schema change, 
+        because the validation might be too complex for a webhook (but this is still TBD). This command could also be used to simulate patches.
```
A similar command to "render" clusters from a class would also be very interesting for creation of classes and CI. That would also allow further validation of patches.

schrej replied:
It's not part of it already imo. That command is only about creating new clusters. I think it would be very useful to be able to statically render a ClusterClass, ideally by just providing all the relevant files. That would make development of ClusterClasses (especially patches) and so on much easier.

And since I have a GitOps perspective: It would allow to statically validate ClusterClasses in a CI pipeline.

sbueringer replied:
Just to confirm. So the command should be able to read a local file with the cluster class and one with a cluster and then output the resulting resources of the reconciliation?

fabriziopandini replied:
I don't think changes are required because IMO this falls under 

> We are considering to implement a clusterctl command to allow the user to simulate the impact of a schema change

And this thread is a +1 to this idea.
Eventually, let's add "; this will be also useful for validation of changes in GitOps pipelines." to keep explicit track of this.

sbueringer replied:
@schrej I think https://github.com/kubernetes-sigs/cluster-api/issues/5319 will address this use case. Feel free to comment on #5319 in case there's something missing

schrej replied:
I agree. Sorry that I didn't respond sooner, I was on vacation the past two weeks.

sbueringer replied:
No problem :). I'll resolve

MarcelMue commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
IMO this line should be changed or removed as this amendment specifically introduces mutable parts to a ClusterClass.

sbueringer replied:
Yeah agree. My problem is that ClusterClass is already mutable even without introducing patches (you can change e.g. templates referenced in the ClusterClass or metadata). 

So the operational benefits (like automatic template rotation) are not patch-specific at all. I'll try to sort this out and bring it in a consistent state.

sbueringer replied:
I tried to clarify it be adding user stories for mutability (independent of patches) and updating the update cluster behavior section.

MarcelMue reviewed (commented):
This makes sense to me _after_ the added context from slack. I personally believe that some more context should be added here to explain the operational benefits which are achieved with the introduction of patches.

CecileRobertMichon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
````diff
@@ -482,6 +796,86 @@ This section talks about updating a cluster which was created using a `ClusterCl
+       - name: region
+         value: us-east-1
+   ```
+   **Note**: `controlPlaneMachineType` will be defaulted to `t3.large` through a mutating webhook.
````
can you expand on this defaulting mechanism? In this case it's defaulted to `t3.large` because that's the default value defined in the example ClusterClass, not because that's an arbitrary default value defined as part of the proposal (might not be obvious to all readers)

CecileRobertMichon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+    - nullable, i.e. each of the basic types can be set to null
+- Basic validation, e.g. format, minimum, maximum, pattern, required, ...
+- Defaulting
+    - Defaulting will be implemented based on the CRD structural schema library and thus will have the same feature set 
```
can we consider allowing built in variables and other variable names as defaults in the future? For example, an Azure vnet name defaults to `{{ .builtin.cluster.name }}-vnet` if not specified.

sbueringer replied:
It's not supported in [JSON Schema](https://json-schema.org/understanding-json-schema/reference/generic.html?highlight=default) nor in [OpenAPI Schema v3](https://swagger.io/docs/specification/data-models/keywords/?sbsearch=default) and thus also not in CRDs.

Additional context:
* the `Default` field in CRDs is a [apiextensionsv1.JSON](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1/types_jsonschema.go#L59) (essentially a byte array)
* the current plan is to re-use the CRD [structural schema defaulting package](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/defaulting) to avoid reimplementing

Independent of that, I think we could implement it, if we want to. A few variants:

#### 1. Constant default values without variables
* could be implemented by just using the CRD defaulting package

#### 2. default values with builtin variables
* we could add a pre-processing step which does go template evaluation over the JSON value
* the result is then used as default value for the CRD defaulting package
Notes:
* Not sure if in this case we should just make the existing `Default` field a template or introduce something like `DefaultTemplate` (which doesn't exist at all in JSON Schema / Open API Schema v3)
* Validation on the ClusterClass wouldn't be entirely trivial, but we could consider something similar to how we would validate variable templates in patches.
* I think we should restrict this to "global" builtin variables which are not "context-sensitive", i.e. `cluster.{name,namespace,labels,annotations}`, `cluster.topology.{version,class}`. As we (currently) only have "global" variables in the Cluster, I think that would make sense.
    * Maybe we should restrict this even further to a subset of those, e.g. labels and annotations don't really make sense as long as we only have basic types, and I'm not sure if we really need them.

#### 3. default values with built-in and user-defined variables

Similar to 2. with the following additional considerations:
* we have to avoid circular dependencies between default values of variables
   * or evaluate them like envsubst: one after another in order of the array and when a referenced variable is not set yet, use the default value
* Validation would be similar as above, but even more brittle

puja108 replied:
Agree, it would be very useful. We're currently using webhooks for such kind of defaulting, but having it built-in would definitely be preferred.

fabriziopandini replied:
FYI, we are just starting to explore external patches, that will be the next evolution of this proposal.
In one of those discussions we considered the idea of allowing external patches to provide additional variables, including their schema and default value.

This has some complexities still to be addressed, but also some clear benefits that can be relevant for this thread:
- External patches will be free to add dynamic defaulting rules, addressing the use case described above (more flexibility)
- From an UX point of view, everything will be simpler. Just add an external patch to a ClusterClass, and then the ClusterClass will get automatically new variables, without the need of manually writing Open API schema definition for them (less yaml)
- Variables will be implicitly aligned to the external patch version (less maintenance)

Generalising a little bit, I personally consider inline patches a no-code, ~simple solution for a limited number of use cases; we can extend it, but we will always hit some limits no matter how much we stretch OpenAPI Schema or JSON Patches standards. Instead, external patches, at least in theory, are a better solution for advanced users, given that they will provide full control on variables/generated patches and the power of real programming languages. 

I'm more than happy to discuss about it if you are interested in this topic

CecileRobertMichon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +139,25 @@ For an end user, the UX to update the kubernetes version of the control plane an
+labels and annotations. This would make it necessary to create new ClusterClasses for small deviations and thus limit the number of variants 
+a single ClusterClass could address.
+
+#### Story 5 - Provide powerful and flexible options to define ClusterClass customizations
```
Would it make sense to add a user story about variable mutations? What happens if I change the value of a variable after clusters using it are created? Are existing clusters updated? If so, in what manner? What about immutable fields?

sbueringer replied:
Added stories for basic mutability (without patches/variables) and one for variable mutation. I also updated the update cluster behavior section to bring it inline with our current implementation (and also to clarify that Cluster mutability is not introduced by patches, we just also leverage it there)

sbueringer commented:
> This makes sense to me _after_ the added context from slack. I personally believe that some more context should be added here to explain the operational benefits which are achieved with the introduction of patches.

Which operational benefits are you referring to specifically?

MarcelMue commented:
> > This makes sense to me _after_ the added context from slack. I personally believe that some more context should be added here to explain the operational benefits which are achieved with the introduction of patches.
> 
> Which operational benefits are you referring to specifically?

Basically the use-case of infra machine templates being mutated / rolled out through this workflow in the long run. This also aligns with the comment from Cecile  about the mutation use case: https://github.com/kubernetes-sigs/cluster-api/pull/5212#discussion_r703783223

sbueringer commented:
> > > This makes sense to me _after_ the added context from slack. I personally believe that some more context should be added here to explain the operational benefits which are achieved with the introduction of patches.
> > 
> > 
> > Which operational benefits are you referring to specifically?
> 
> Basically the use-case of infra machine templates being mutated / rolled out through this workflow in the long run. This also aligns with the comment from Cecile about the mutation use case: [#5212 (comment)](https://github.com/kubernetes-sigs/cluster-api/pull/5212#discussion_r703783223)

Thx, that's helpful. Will try to make that clear and especially the mutability we already have without patches and how patches relate to that.

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -91,6 +100,7 @@ This method of provisioning the cluster would act as a single control point for
 - Extend the Cluster object to use ClusterClass for creating managed topologies.
 - Enhance the Cluster object to act as a single point of control for the topology.
 - Extend the Cluster controller to create/update/delete managed topologies.
+- Introduce mechanisms for allowing Cluster-specific transformations of a ClusterClass, and the continuous reconciliation of the Cluster managed resources.
```
```suggestion
- Introduce mechanisms for allowing Cluster-specific transformations of a ClusterClass, and the continuous reconciliation of the ClusterClass managed resources.
```
Is this more correct? Or Cluster's managed resources?

sbueringer replied:
Hm maybe it's rather `of the topology managed resources`

killianmuldoon replied:
Yeah - that works I think.

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,34 +532,124 @@ type LocalObjectTemplate struct {
   - `spec.workers.machineDeployments[i].class` field must be unique within a ClusterClass.
+  - `VariableDefinitionClass`:
+    - names must be unique, not empty and not equal to `builtin`
+    - schemas must be valid
```
Valid OpenAPI v3 schemas?

sbueringer replied:
It's not the same as a OpenAPI v3 schema (not even the one in CRDs matches the spec entirely). 
I would prefer keeping this a bit vague in the proposal as the exact validation is very close to what is done for CRDs and rather complex.

killianmuldoon replied:
Sure - understood. I was just hoping there was a simple answer to this complicated question 😄

sbueringer replied:
Unfortunately, there will probably only be one rather long answer in the PR when we implement it :)

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -482,6 +796,86 @@ This section talks about updating a cluster which was created using a `ClusterCl
+     - evaluate patch selector
+     - evaluate patch values
+   - Apply patches to our local copies of the templates. 
+     <br>**Note**: Patches are applied in the order in which they are defined in the ClusterClass.
```
Is this an implementation detail or reliable API behaviour?

vincepri replied:
+1 to clarify, it should be an API behavior

sbueringer replied:
I added this note here as additional information. I also have the same note 3x in the godoc comments at the places where it's relevant. Is there another place where it's more appropriate to put this information?

killianmuldoon replied:
I think the behavior is pretty clear here - I just wanted to be sure this is the intended behaviour rather than just the behavior as currently implemented. Intended behaviour is the assumption I would have had reading this doc.

sbueringer replied:
Ah got it. Yes it is intended (there's no implementation yet).

@vincepri do you also think it's okay as is or should I make it more explicit?

sbueringer replied:
I'll assume it's clearn enough

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -229,8 +262,182 @@ type LocalObjectTemplate struct {
+type JSONPatchDefinitionValue struct {
+  // Variable is the variable to be used as value.
+  // Variable can be one of the variables defined in .spec.variables or a builtin variable.
+  // Note: Builtin variables are described in the CAPI book.
```
Is this useful context? This isn't in the book right now, right? Won't this and a lot more end up in the book as documentation eventually?

sbueringer replied:
I can drop it, but I thought otherwise I would get comments where builtin variables are defined. (as they are not defined here in the struct section of the proposal)

Also eventually we seem to try to keep the proposal structs more or less sync with the ones we have eventually in the code. So I though it's a good idea to add it.

killianmuldoon replied:
I think the best solution is either to describe them here or to leave them undefined - this way splits the difference in a way that doesn't seem to add anything to the proposal.

sbueringer replied:
Okay, I'll drop it

sbueringer commented:
@enxebre @schrej @MarcelMue @CecileRobertMichon @killianmuldoon @vincepri @fabriziopandini

Thx for your reviews, ptal.

I tried to address the comments. The biggest change is that I tried to clarify how Cluster mutability works (even before introducing patches).

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
nit - just a grammar preference, but these unnecessary commas make things harder to parse IMO, even though a bunch of automated tools seem to love them 😐 
```suggestion
However we are intentionally leaving them out from this initial iteration for the following reasons:
```

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +138,53 @@ For an end user, the UX to update the kubernetes version of the control plane an
+- Changes to InfrastructureMachineTemplates and BootstrapTemplates should be rolled out according to the established operational practices documented in 
+[Updating Machine Infrastructure and Bootstrap Templates](https://cluster-api.sigs.k8s.io/tasks/updating-machine-templates.html), i.e. "template rotation".
+- There are provider-specific incompatible changes which cannot be validated in a "core" webhook, e.g. changing an immutable field of `KubeadmControlPlane`. Those changes will inevitably lead to
+  errors during topology reconciliation. Those errors should be surfaced on the Cluster resource. To mitigate this issue, we should consider implementing a clusterctl command to simulate changes like 
```
```suggestion
  errors during topology reconciliation. Those errors should be surfaced on the Cluster resource. 
```
Maybe include the clusterctl idea in 'future work' or directly as an issue. It becomes a weird implementation thought here, not tracked and not clear.

sbueringer replied:
@fabriziopandini WDYT?

sbueringer replied:
Ah thought it's an another location. I'll just drop it. We have to revisit the clusterctl additions anyway.

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +138,53 @@ For an end user, the UX to update the kubernetes version of the control plane an
+  this locally.
+
+#### Story 5 - Ability to customize individual clusters via variables
+As an end user, I want to customize individual clusters simply by providing variables in the Cluster resource.
```
```suggestion
As an end user, I want to customize individual Clusters simply by providing variables in the Cluster resource.
```

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +138,53 @@ For an end user, the UX to update the kubernetes version of the control plane an
+It should be possible to define simple customizations inline in a ClusterClass without writing any code. To support more advanced use cases, there should be an external mechanism which allows to leverage 
+the full power of a programming language.
+
+**Note**: The external mechanism will be addressed in another iteration of this proposal.
```
Maybe better to move this up to future work?

sbueringer replied:
I'll add it to future work, but I want to also keep it here, because I also write about how we need something easy like inline patches *and* something more complex and then never mention the external mechanism later in the doc.

killianmuldoon commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +556,125 @@ type LocalObjectTemplate struct {
+        not compatible. We are considering to implement a clusterctl command to allow the user to simulate the impact of a schema change, 
+        because the validation might be too complex for a webhook (but this is still TBD). This command could also be used to simulate patches.
+    - variables cannot be removed as long as they are used in Clusters
+      - We are considering to add a field to allow deprecations of variables. When a variable would be deprecated no new usages of 
```
```suggestion
      - We are considering adding a field to allow deprecations of variables. When a variable would be deprecated no new usages of 
```
nit

fabriziopandini commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +557,125 @@ type LocalObjectTemplate struct {
+It’s also possible to use so-called builtin variables in addition to user-defined variables. The following builtin variables are available:
+- `cluster.{name,namespace,labels,annotations}`
+- `cluster.topology.{version,class}`
+- `cluster.topology.controlPlane.{labels,annotations,replicas}`
```
```suggestion
- `cluster.topology.controlPlane.{labels,annotations,replicas,version}`
```
Adding version given that it could different that the topology one

fabriziopandini commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +557,125 @@ type LocalObjectTemplate struct {
+- `cluster.topology.{version,class}`
+- `cluster.topology.controlPlane.{labels,annotations,replicas}`
+    - **Note**: these variables are only available when patching control plane or control plane machine templates.
+- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,class,name}`
```
```suggestion
- `cluster.topology.machineDeployment.current.{labels,annotations,replicas,version,class}`
```
Adding version given that it could different that the topology one; removing name given that we apply patches to templates in MachineDeploymentClasses (and then the class should be the same for all the MachineDeployment instances we derive from it).

sbueringer commented:
Hey folks,
I think I addressed all comments. I would like to propose to start a lazy consensus if there are no concerns.

It would be great to get start with the implementation and iterate on it. It would also make it a lot easier for our team to coordinate the work to get v1beta1 and ClusterClass patches done.

I'll also bring this up in the meeting today.

vincepri commented:
/milestone v1.0

joejulian commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +140,52 @@ For an end user, the UX to update the kubernetes version of the control plane an
+**Notes**: Same notes as in Story 4 apply.
+
+#### Story 7 - Provide powerful and flexible options to define ClusterClass customizations
+As a ClusterClass author, I want to have powerful and flexible options to specify how templates referenced in a ClusterClass can be customized through patches. These patches should use the user-provided 
+variables defined in a Cluster as input. A ClusterClass author should be able to abstract away details of the provider configuration and thus simplify cluster configuration for the end user.
+
+It should be possible to define simple customizations inline in a ClusterClass without writing any code. To support more advanced use cases, there should be an external mechanism which allows to leverage 
+the full power of a programming language.
```
I don't understand "powerful and flexible" in this section and it becomes even more confusing as I read about simplicity and "without writing any code". It's conflicting and confusing.

Perhaps it could be rephrased as a story that states who is trying to do what, maybe? This reads more like a goal to me.

sbueringer replied:
I rewrote the whole story and moved it up.

joejulian commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +140,52 @@ For an end user, the UX to update the kubernetes version of the control plane an
+  errors during topology reconciliation. Those errors should be surfaced on the Cluster resource.
+
+#### Story 5 - Ability to customize individual Clusters via variables
+As an end user, I want to customize individual Clusters simply by providing variables in the Cluster resource.
+
+**Note**: Without this ability all Clusters of the same ClusterClass would be the same apart from the properties that are already configure via the topology,
+like Kubernetes version, labels and annotations. This would limit the number of variants a single ClusterClass could address. Variables can e.g. be used to make 
+region/location, machine types and network configuration configurable.
+
+**Example**: A user wants to deploy CAPA Clusters using ClusterClass in different AWS regions. One option to achieve this is to duplicate the ClusterClass and its referenced templates.
+The better option is to introduce a variable and a corresponding patch in the ClusterClass. Now, a user can simply set the AWS region via a variable in the Cluster spec, instead of 
+having to duplicate the entire ClusterClass just to set a different region in the AWSCluster resource.
```
The principal of least surprise suggests that the opposite is true, that having duplicate, specific, ClusterClass resources provides an accurate reflection of the desired state and is therefore the better option. This allows for auditing and history and reduces the impact of human error.

schrej replied:
Since there aren't any condition based variables or side effects proposed, the risk of surprises is pretty low, isn't it? Without variables, cluster classes are  useless in our environment for example, since we would have to create one per cluster.

sbueringer replied:
I think both use cases are valid. Some prefer having one ClusterClass per cluster others one ClusterClass for as many Clusters as possible. I think it highly depends on the use case and what automation is used around CAPI (e.g. GitOps with plain YAML files or another "higher-level" Operator). It's also a big difference if only a few clusters will be managed or hundreds/thousands. 

In my opinion the first use case is already supported without patches, but we need patches to make the second approach viable. Both will be supported at the end.

joejulian commented:
I don't understand this at all. Kubernetes is a declarative state resolution engine. This is not state declaration, it's templating which doesn't belong in a declarative resource. There are already tools for creating templated resources and expectations that users will use those tools for creating simple outputs.

Wouldn't it be more aligned with the rest of the community if we just provided helm and/or kustomize templates to create sets of ClusterClass resources?

vincepri commented:
Thanks for the feedback @joejulian!

> I don't understand this at all. Kubernetes is a declarative state resolution engine. This is not state declaration, it's templating which doesn't belong in a declarative resource.

Absolute statements aside, this proposal tackles the problem that declarative specs to manage Kubernetes clusters are incredibly dynamic and have lots of intersections with each other. 

A Cluster API Cluster won't become a Kubernetes cluster unless multiple objects cooperate and are safely configured, as a brief example KubeadmConfig (bootstrap) has to align with its infrastructure part, which has to then align to control plane options, which also has to align with network configurations, and so on.

The N-dimensional matrix of options we have today is a powerful tool which brings solid foundations. The user experience isn't ideal and we don't have a way to safely stamp Clusters and reuse template resources across many clusters.

> Wouldn't it be more aligned with the rest of the community if we just provided helm and/or kustomize templates to create sets of ClusterClass resources?

We're already allowing any external tools like helm, kustomize, ytt to create Cluster API clusters, the functionality won't go away and this proposal doesn't impose any restrictions or replacement for those templating engines.

That said, these templating tools cannot access or validate resources properly until a cluster is reconciled by the N controllers involved.

The project's goal is to make Kubernetes cluster lifecycle boring, easy, validated, and safe. In the past 3 years we've been collecting feedback from both users, developers, and cluster operators which brought us here where ClusterClass and managed topologies is taking a step further in creating safe declarative dynamic patterns that can be fully validated when a cluster is created or updated.

dlipovetsky commented:
> A Cluster API Cluster won't become a Kubernetes cluster unless multiple objects cooperate and are safely configured, as a brief example KubeadmConfig (bootstrap) has to align with its infrastructure part, which has to then align to control plane options, which also has to align with network configurations, and so on.

As a community, we can (and some of us do) build tooling that generates a complete and nominally valid set of objects for a cluster. Such tooling composes with the existing Cluster API objects. 

> That said, these templating tools cannot access or validate resources properly until a cluster is reconciled by the N controllers involved.

To get a complete, nominally valid set of objects for a cluster, you generate the objects, then submit them using server-side dry-run. (Even then, you may be unable to create all the objects, e.g. after the dry-run returns, another client creates an object whose name conflicts with some object in your set. Kubernetes does not support transactions. You can work around this, as clusterctl move does, by "pausing" the objects, then creating them, then "unpausing" them).

After the server-side dry-run, you _have_ to let the controllers reconcile in order to find out whether the objects are in fact valid. For example, CAPA validation webhooks won't tell you whether the AMI ID is correct for the Kubernetes version you specify, or whether the AMI ID even exists. 
 
If I understand correctly, the logical inverse of the claim here is: "ClusterClass can access and validate resources properly before the cluster is reconciled by the N controllers." I would like to understand how ClusterClass achieves this, if tooling that composes with existing Cluster API objects cannot.

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +140,52 @@ For an end user, the UX to update the kubernetes version of the control plane an
 - Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
 
+#### Story 4 - Use ClusterClass to easily modify Clusters
+As an end user, I want to be able to easily change the configuration of all Clusters of a `ClusterClass`. For example, I want to be able to change the ControlPlane configuration by  
```
can we be more specific about "end user", would it be a cluster admin in this scenario?

sbueringer replied:
Looking at some of our other proposals, we seem to be calling this persona "cluster operator". I used "end user" because the 3 pre-existing user stories used it. I"m fine with consistently changing it to "cluster operator"

sbueringer replied:
Done (I'll push later)

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +140,52 @@ For an end user, the UX to update the kubernetes version of the control plane an
 
+#### Story 4 - Use ClusterClass to easily modify Clusters
+As an end user, I want to be able to easily change the configuration of all Clusters of a `ClusterClass`. For example, I want to be able to change the ControlPlane configuration by  
+changing the ControlPlane ref to point to another ControlPlane with a different configuration. The same should be possible for all templates referenced in a ClusterClass.
```
>For example, I want to be able to change the ControlPlane configuration by changing the ControlPlane ref to point to another ControlPlane with a different configuration. The same should be possible for all templates referenced in a ClusterClass.

I feel this is describing a viable technical behaviour, can we try to elaborate why would one want to make use of this behaviour, what's the real world user story/scenario where this technical feature helps?

sbueringer replied:
You're right. I should have left the details how the config is changed out of it.

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -129,13 +140,52 @@ For an end user, the UX to update the kubernetes version of the control plane an
+having to duplicate the entire ClusterClass just to set a different region in the AWSCluster resource.
+
+#### Story 6 - Ability to mutate variables
+As an end user, I want to be able to modify variables in a Cluster. The variable modification should trigger a reconciliation which rolls out the managed topology accordingly, i.e. 
```
Similar to https://github.com/kubernetes-sigs/cluster-api/pull/5212/files#r712039251
I feel this is more of a technical ability description that could may be conflated with story 4 or 5 to support a real world scenario which is the actual user story?

sbueringer replied:
I'll keep the stories separate (story 4 talks about general mutability without considering patches at all, story 5 already has a big scope). I'll add an example and drop the technical detail about patching

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -90,16 +99,18 @@ This method of provisioning the cluster would act as a single control point for
 - Enhance the Cluster object to act as a single point of control for the topology.
-- Extend the Cluster controller to create/update/delete managed topologies.
+- Extend the Cluster controller to create/update/delete managed topologies (this includes continuous reconciliation of the topology managed resources).
+- Introduce mechanisms to allow Cluster-specific customizations of a ClusterClass 
```
In the Glossary, clusterClass
>A collection of templates that define a topology (control plane and machine deployments) to be used to create one or more clusters.

Should we replace with

>A collection of templates that define a topology (control plane and machine deployments) that continuously reconcile one or more clusters

or something similar?

sbueringer replied:
Done

sbueringer commented:
@enxebre Thx for your review, findings should be adressed, PTAL :)

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -115,27 +125,66 @@ This proposal enhances the `Cluster` object to create topologies using the `Clus
+As a cluster operator, I want to be able to easily scale up/down the number of replicas for each set of worker nodes in the cluster.
 - Currently, (for a cluster with 3 machine deployments) this is possible by updating these three different objects representing the sets of worker nodes in the pool. An easier user experience would be to update a single object to enable the scaling of multiple sets of worker nodes.
 
+#### Story 4 - Use ClusterClass to easily modify Clusters
```
nit: Use ClusterClass to easily modify Clusters in bulk?

enxebre commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -115,27 +125,66 @@ This proposal enhances the `Cluster` object to create topologies using the `Clus
+As a ClusterClass author (e.g. an infrastructure provider author), I want to be able to write a ClusterClass which covers a wide range of use cases. To make this possible, 
+I want to make the CusterClass customizable, i.e. depending on configuration provided during Cluster creation, the managed topology should have a different shape.
+
+**Note**: Without this feature all Clusters of the same ClusterClass would be the same apart from the properties that are already configure via the topology,
```
can we have an example here?

sbueringer replied:
Done

enxebre commented:
@sbueringer I'm in favour of moving along with this so we can get real use feedback and then adapt appropriately. I'm still navigating through the impl details but other than some natural concern about introducing complexity/degrade UX this lgtm.

MarcelMue commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +559,125 @@ type LocalObjectTemplate struct {
+     valueFrom:
+       template: |
+         - contentFrom:
+             secret
```
```suggestion
             secret:
```
I assume this needs to be valid yaml as well.

sbueringer replied:
Yes, thx

MarcelMue commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +559,125 @@ type LocalObjectTemplate struct {
+    - names must be unique and not empty
+  - `PatchDefinition`:
+    - selector must have valid field values and match at least one template
+    - jsonPatches:
```
The only problem I see with jsonPatches is the lack of list iteration in JSONPatch. AFAIK it is not possible to change e.g. every item in a list in a single JSONPatch. The length of the list must be known in order to do it.

I don't think that's a blocker as JSONPatch is widely used but something to be aware of (if any usecases for ClusterClass include patching multiple elements in lists).

fabriziopandini replied:
agreed this is not a blocker, also this https://github.com/kubernetes-sigs/cluster-api/pull/5212#discussion_r704184935 could provide a powerful iteration

MarcelMue reviewed (commented):
Did the discussion on topics shared by @dlipovetsky take place anywhere else? I can not see a resolution in this PR.

sbueringer commented on docs/proposals/202105256-cluster-class-and-managed-topologies.md:
```diff
@@ -313,35 +559,125 @@ type LocalObjectTemplate struct {
+    - schemas are mutable
+      - The current assumption is that we validate schema changes against existing clusters and block in case the changes are 
+        not compatible. We are considering to implement a clusterctl command to allow the user to simulate the impact of schema and patch changes, 
+        because the validation might be too complex for a webhook (but this is still TBD); this will be also useful for validation of changes in GitOps pipelines.
```
I think https://github.com/kubernetes-sigs/cluster-api/issues/5319 addresses the clusterctl part of it.

In theory we want as much validation as possible in webhooks. What we don't want to do in the webhook is:
* essentially run the whole reconcile loop in a dry-run mode to calculate the new resources of the managed topology (KCP, MD, ...)
* take the resulting resources of this dry-run and execute dry-run create/updates on other resources to test if their webhook would accept the change.

fabriziopandini replied:
my two cents on this.
this is not technically doable with the default web hooks provided by controller runtime as of today.
we are exploring with to controller runtime folks if/how to support more powerful webhook so making it possible a clean solution, but being this a WIP we don't have full control on, I think that it is reasonable to keep all the options on the table in the proposal (still TBD).
In other words, it is required to make some concrete work here before committing to a final solution.

sbueringer commented:
> Did the discussion on topics shared by @dlipovetsky take place anywhere else? I can not see a resolution in this PR.

Not that I'm aware of.

I'm not sure how I can address this discussion in the proposal itself. To me it seems like there is a difference in opinion if ClusterAPI - as it is today - is already enough and users of ClusterAPI should build be forced to build on top of ClusterAPI with tools like helm/kustomize/ytt/... . I don't think **we** should build a multi-cluster management layer **with those tools**.

In my opinion the addition of ClusterClass (and ClusterClass patches to make a ClusterClass usable for more than one Cluster) is a huge UX improvement which should be part of the project. But of course we should preserve the current APIs to make it possible to build on top of ClusterAPI without being forced to use ClusterClass.

MarcelMue commented:
/lgtm

fabriziopandini commented:
/lgtm

vincepri commented:
@sbueringer please squash

sbueringer commented:
> @sbueringer please squash

Done

fabriziopandini commented:
/lgtm

MarcelMue commented:
/lgtm

vincepri commented:
/approve
/hold for community meeting

sbueringer commented:
As discussed in the CAPI meeting today, lazy consensus until ~Friday
